initial fix

momolarson opened the pull request:
First pass at fixing issue 911

sschmidTU commented:
did you forget to push a commit? because it seemed like you wanted to average the undefined skyline values between the previous and next value, as we said in #911.
But you don't necessarily need to write this yourself, i can also do it a bit later today.

sschmidTU commented:
Or maybe we just copy the previous value, because it would be a bit complicated to look ahead in the values, and this may be the better solution anyways

momolarson commented:
yeah, I was thinking the same.  I have both solutions here, but I agree with the copy previous value.  They are both in current PR.

sschmidTU commented:
Or rather, let's take the max value, because we could have something like [undefined, 95, 95], or [0,0,undefined,95].

sschmidTU commented on src/MusicalScore/Graphical/SkyBottomLineCalculator.ts:
```diff
@@ -152,6 +164,41 @@ export class SkyBottomLineCalculator {
+                return 0;
+            }
+        }
+    }
```
we need to return 0 after the for loop as well, in case no value was not undefined

sschmidTU commented on src/MusicalScore/Graphical/SkyBottomLineCalculator.ts:
```diff
@@ -152,6 +164,41 @@ export class SkyBottomLineCalculator {
+     */
+    private findPreviousValidNumber(start: number, tSkyLine: []): number {
+        for (let idx: number = start; idx >= 0; idx--) {
+            if (!isNaN(tSkyLine[idx])) {
```
this should be !isNan && !== undefined

sschmidTU replied:
oh, nevermind, isNaN(undefined) is true, so this works.

sschmidTU commented on src/MusicalScore/Graphical/SkyBottomLineCalculator.ts:
```diff
@@ -152,6 +164,41 @@ export class SkyBottomLineCalculator {
+                return tSkyLine[idx];
+            }
+        }
+    }
```
here return 0 after the for loop as well

sschmidTU commented on src/MusicalScore/Graphical/SkyBottomLineCalculator.ts:
```diff
@@ -152,6 +164,41 @@ export class SkyBottomLineCalculator {
+     */
+    private findNextValidNumber(start: number, tSkyLine: Array<number>): number {
+        if (start === tSkyLine.length) {
+            return tSkyLine[start];
```
this won't work, tSkyLine[tSkyLine.length] is an index out of bounds error.

sschmidTU commented:
i made a few review comments, but i can simply merge this on a branch and polish the last few things myself if you want.

sschmidTU commented on src/MusicalScore/Graphical/SkyBottomLineCalculator.ts:
```diff
@@ -190,7 +190,7 @@ export class SkyBottomLineCalculator {
      */
     private findNextValidNumber(start: number, tSkyLine: Array<number>): number {
         if (start === tSkyLine.length) {
```
we could also do >= to be safe on values > tSkyLine.length.

sschmidTU commented on src/MusicalScore/Graphical/SkyBottomLineCalculator.ts:
```diff
@@ -190,7 +190,7 @@ export class SkyBottomLineCalculator {
     private findNextValidNumber(start: number, tSkyLine: Array<number>): number {
         if (start === tSkyLine.length) {
-            return tSkyLine[start];
+            return tSkyLine[start - 1];
```
good!

sschmidTU commented on src/MusicalScore/Graphical/SkyBottomLineCalculator.ts:
```diff
@@ -101,7 +101,19 @@ export class SkyBottomLineCalculator {
+
+            for (let idx: number = 0; idx < tmpSkyLine.length; idx++) {
+                if (tmpSkyLine[idx] === undefined) {
+                    tmpSkyLine[idx] = (this.findPreviousValidNumber(idx, tmpSkyLine) + this.findNextValidNumber(idx, tmpSkyLine)) / 2; // tmpSkyLine[x-1];
```
let's do Math.max(previous, next) here instead of the average

sschmidTU commented on src/MusicalScore/Graphical/SkyBottomLineCalculator.ts:
```diff
@@ -152,6 +164,43 @@ export class SkyBottomLineCalculator {
+     */
+    private findPreviousValidNumber(start: number, tSkyLine: number[]): number {
+        for (let idx: number = start; idx >= 0; idx--) {
+            if (!isNaN(tSkyLine[idx]) && tSkyLine[idx] !== undefined) {
```
we don't need the `&& tSkyLine[idx] !== undefined`, sorry, that was my mistake. isNaN() covers undefined.

momolarson replied:
all good!  I appreciate your review!

sschmidTU commented on src/MusicalScore/Graphical/SkyBottomLineCalculator.ts:
```diff
@@ -152,6 +164,43 @@ export class SkyBottomLineCalculator {
+        for (let idx: number = start; idx >= 0; idx--) {
+            if (!isNaN(tSkyLine[idx]) && tSkyLine[idx] !== undefined) {
+                return tSkyLine[idx];
+            } else {
```
remove the else branch. this would always return 0 if the first previous value is undefined. but we want to keep looking until we find a better previous value.
we now return 0 after the for loop, that's why we don't need the else anymore.

sschmidTU replied:
This is already fixed in findNextValidNumber, it looks like.

sschmidTU commented:
@momolarson great, this looks good to me now! Are you done with your changes? I'll just run the tests and merge then.

momolarson commented:
> @momolarson great, this looks good to me now! Are you done with your changes? I'll just run the tests and merge then.

yeah, I believe I have addressed all of your issues.  thanks for the help.

sschmidTU commented:
Alright, thanks for the fix! Merging.
